feat(benchmarks): Add Prefix Caching Benchmark to Serving Benchmark

ywang96 opened the pull request:
Reopens #3194 that was closed due to fork cleanup

Additional changes not mentioned in #3194 :
- Feature to save individual token latencies (itl) per request to the result json for further debugging
- Default `vllm` benchmark to OpenAI Completion API
- ~~If streaming is supported, count output tokens by counting actual token streaming SSE events instead of tokenization~~ will apply tokenization on output to count tokens since some API server does not stream token by token.
- Misc fixes/workaround in case the API server doesn't strictly follow Open API

To run the benchmark on the sonnet dataset, specify `--dataset-name sonnet` and `--dataset-path <path to sonnet.txt>`. Lengths of input, output and prefix lengths can be specified with command line args.

ywang96 committed:
recommit

ywang96 committed:
comment

ywang96 committed:
yapf

ywang96 committed:
assert chat model on prefix benchmark

ywang96 committed:
Merge branch 'vllm-project:main' into prefix-benchmark

ywang96 committed:
add request-level token latencies

ywang96 committed:
Merge branch 'prefix-benchmark' of https://github.com/ywang96/vllm into prefix-benchmark

ywang96 committed:
Merge branch 'main' into prefix-benchmark

ywang96 committed:
typo

ywang96 committed:
yapf

ywang96 committed:
remove generation prompt

ywang96 committed:
revert

ywang96 committed:
iterate

ywang96 committed:
Merge branch 'main' into prefix-benchmark

ywang96 committed:
format

ywang96 committed:
typo & comment

ywang96 committed:
deprecate vllm generate api server

ywang96 committed:
add optional args for recordkeeping

ywang96 committed:
save deployment spec to json

ywang96 committed:
workaround for non-streaming

ywang96 committed:
fix ds payload

ywang96 committed:
fix ds payload

ywang96 committed:
iterate

ywang96 committed:
typo

ywang96 committed:
Merge branch 'main' into prefix-benchmark

ywang96 committed:
format

ywang96 committed:
yapf

ywang96 commented:
@simon-mo Please take a look whenever you're free

cc @robertgshaw2-neuralmagic in case you want to use this version of the server benchmark script - feel free to review it as well, Thanks!

ywang96 committed:
update script

ywang96 committed:
update comment

ElizaWszola commented on benchmarks/benchmark_serving.py:
```diff
@@ -326,10 +479,17 @@ def main(args: argparse.Namespace):
+        "--dataset-name",
+        type=str,
+        default="sharegpt",
+        choices=["sharegpt", "sonnet"],
```
Would it make sense to, rather than getting dataset by name, specify whether we want to read from json or from a text file and provide a path to it? Or is it an overkill?

ywang96 replied:
Hmm - I think we can add the option to read from a json or path and that'll be nice for sure, but the problem is we'll have to use a single user base prompt for all datasets. Also certain arguments of the benchmark (e.g., configure input, output & prefix length) is only used for certain dataset. 

I haven't come up with a good solution so for now I keep them as separate datasets. I really like the idea of data registry where the way we sample prompts & output from the datasets lives outside the main benchmark script, and we can reuse it for other benchmark scripts as well, but I haven't put too much thoughts into it

ywang96 committed:
iterate

ywang96 committed:
readability

ywang96 committed:
format

simon-mo commented on benchmarks/benchmark_serving.py:
```diff
@@ -313,7 +450,29 @@ def main(args: argparse.Namespace):
+        help="Version of the serving backend/engine, if known. "
+        "Only used in result json for record keeping purposes.",
+    )
+    parser.add_argument(
+        "--tp-size",
+        type=str,
+        default="N/A",
+        help="Size of tensor parallelism of the model server, if known. "
+        "Only used in result json for record keeping purposes.",
+    )
+    parser.add_argument(
+        "--pp-size",
+        type=str,
+        default="N/A",
+        help="Size of pipeline parallelism of the model server, if known. "
+        "Only used in result json for record keeping purposes.",
+    )
+    parser.add_argument(
+        "--device-name",
+        type=str,
+        default="N/A",
+        help="Device name, if known. Only used in result json for record "
+        "keeping purposes.",
     )
```
merge these into just generic kv pairs for metadata? `--metadata pp-size=2 --metadata tp-size=4`.

ywang96 replied:
Good point! I will separate metadata from the those actually matter to the launch of the benchmark

ywang96 replied:
Updated! Now metadata can be specified like `--metadata tp=2 version=0.3.3 device-name=A100-80G`, and it will be only used if the end user is saving the benchmark result as json file.

simon-mo commented on benchmarks/benchmark_serving.py:
```diff
@@ -359,6 +525,27 @@ def main(args: argparse.Namespace):
         default=1000,
         help="Number of prompts to process.",
     )
+    parser.add_argument(
+        "--input-len",
+        type=int,
+        default=550,
+        help=
+        "Number of input tokens per request, used only for sonnet dataset.",
+    )
+    parser.add_argument(
+        "--output-len",
+        type=int,
+        default=150,
+        help=
+        "Number of output tokens per request, used only for sonnet dataset.",
+    )
+    parser.add_argument(
+        "--prefix-len",
+        type=int,
+        default=200,
+        help=
+        "Number of prefix tokens per request, used only for sonnet dataset.",
+    )
```
name them --sonnet-input-len, ... etc because the names are so generic that I can imagine anyone will misuse them

ywang96 committed:
move metadata

ywang96 committed:
iterate

ywang96 committed:
iterate

ywang96 committed:
Merge branch 'main' into prefix-benchmark

ywang96 commented:
@simon-mo I've addressed your comments as well as adding `--dataset` back so it's backward compatible for now with a deprecation warning. PTAL, thanks!

ywang96 committed:
add stack level

ywang96 committed:
clarify

ywang96 committed:
yapf & ruff

ywang96 committed:
isort

dgisser commented:
@ywang96 is there any plan to post the metrics you've calculated based on these benchmark scripts (or the slides from the vllm meetup)?

ywang96 commented:
> @ywang96 is there any plan to post the metrics you've calculated based on these benchmark scripts (or the slides from the vllm meetup)?

There is and stay tuned!
